### PR TITLE
0.6.0-rc1 Clusterctl Omnibus PR

### LINKED ISSUE #753: provide workload cluster template for clusterctlv2
/kind feature

**Describe the solution you'd like**
CAPV should provide provide workload cluster templates for clusterctl. The 2 flavours that we support are the following:

- mono: mono-instance control plane relying on `kubeadmConfig`
- ha: HA control plane relying on HA Proxy as a load balancer and the `KubeadmControlPlane`

/priority important-soon
/assign @randomvariable 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-vsphere version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### LINKED ISSUE #774: Validate / fixup manager defaults
/kind bug

**What steps did you take and what happened:**
By default, all instances of capv will attempt leader election in the capv-system namespace. For v1alpha3, this breaks the webhooks

Fix up manifests / and or sync with CAPA in terms of behaviour.

Blocks #753 

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-vsphere version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by randomvariable**
**What this PR does / why we need it**:
Introduces support for clusterctl
The number of templated variables has been reduced to a minimum, so as to minimise the amount to get started. Production users should always modify resultant outputs from cluster

Additionally 
* Fixes issue with loadbalancer controller and owner references.
* New dev workflow for clusterctl, similar to the AWS provider
* Support for tilt
* Small fix for linters for developers using Linux with SELinux enabled, (sideported from k/k)
* Fix to readme that incorrectly noted 1.17.4 instead of 1.17.3 OVAs

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #753 
Fixes #774 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

TODO:
* Create issue to merge with e2e generators as there's overlap there (out of scope for this PR)
* Create issue around event record pod name

**Release note**:

```release-note
* Added support for clusterctlv2, with new cluster template for Cluster API v1alpha3
* [development]: Added tilt configuration for rapid development
* HAProxyLoadBalancer controller will always look up a vSphereCluster in the same namespace
* [development]: Linters work on SELinux enabled systems
* Update links to OVAs
* Running CAPV will no longer run the webhook server by default
* The `--pod-namespace` flag has been removed. Leader election will always look up a configmap in the same namespace in which CAPV runs.
* A `--enable-leader-election` flag has been added to support the deployment of webhooks as a separate instance.
```

### CODE REVIEW

**Comment by randomvariable** [comment]
i'll sort out lints

File: examples/default/addons.yaml [hunk 1]
```diff
@@ -623,7 +623,7 @@ spec:
             # chosen from this range. Changing this value after installation will have
             # no effect. This should fall within `--cluster-cidr`.
             - name: CALICO_IPV4POOL_CIDR
-              value: "${CLUSTER_CIDR}"
+              value: "192.168.0.0/16"
```
Comment (yastij) [inline examples/default/addons.yaml]: why is this being hardcoded to `192.168.0.0/16` ?

File: main.go [hunk 2]
```diff
@@ -81,7 +56,7 @@ func main() {
 	flag.StringVar(
 		&managerOpts.MetricsAddr,
 		"metrics-addr",
-		":8084",
+		":8080",
```
Comment (yastij) [inline main.go]: we should document this change on the release-note

File: docs/getting_started.md [hunk 3]
```diff
@@ -48,27 +48,27 @@ In order for `clusterctl` to bootstrap a management cluster on vSphere, it must
 
 It is required that machines provisioned by CAPV use one of the official CAPV machine images as a VM template. The machine images are retrievable from public URLs. CAPV currently supports machine images based on Ubuntu 18.04 and CentOS 7. A list of published machine images is available [here](../README.md#kubernetes-versions-with-published-ovas). For this guide we'll be deploying Kubernetes v1.16.2 on Ubuntu 18.04 (link to [machine image](https://storage.googleapis.com/capv-images/release/v1.16.2/ubuntu-1804-kube-v1.16.2.ova)).
 
-[Create a VM template](https://docs.vmware.com/en/VMware-vSphere/6.7/com.vmware.vsphere.vm_admin.doc/GUID-17BEDA21-43F6-41F4-8FB2-E01D275FE9B4.html) using the OVA URL above. The rest of the guide will assume you named the VM template `ubuntu-1804-kube-v1.16.2`.
+[Create a VM template](https://docs.vmware.com/en/VMware-vSphere/6.7/com.vmware.vsphere.vm_admin.doc/GUID-17BEDA21-43F6-41F4-8FB2-E01D275FE9B4.html) using the OVA URL above. The rest of the guide will assume you named the VM template `ubuntu-1804-kube-v1.16.3`.
```
Comment (yastij) [inline docs/getting_started.md]: 1.16.7 is latest patch one.

File: packaging/flavorgen/flavors/generators.go [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+				Workspace: infrav1.CPIWorkspaceConfig{
+					Server:       vSphereServerVar,
+					Datacenter:   vSphereDataCenterVar,
+					Datastore:    vSphereDatastoreVar,
+					ResourcePool: resourcePoolVar,
+					Folder:       folderVar,
+				},
+				ProviderConfig: infrav1.CPIProviderConfig{
+					Cloud: &infrav1.CPICloudConfig{
+						ControllerImage: defaultCloudProviderImage,
```
Comment (yastij) [inline packaging/flavorgen/flavors/generators.go]: shouldn't be this accessed through `cloudprovidersvc` too ? (i.e. we should add it if not)

File: packaging/flavorgen/flavors/generators.go [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+*/
+
+package flavors
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	infrav1 "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3"
+	cloudproviderSvc "sigs.k8s.io/cluster-api-provider-vsphere/pkg/services/cloudprovider"
```
Comment (yastij) [inline packaging/flavorgen/flavors/generators.go]: `cloudprovidersvc`

File: packaging/flavorgen/flavors/generators.go [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+	folderVar                   = "${ VSPHERE_FOLDER_VAR }"
+	machineDeploymentNameSuffix = "-md-0"
+	resourcePoolVar             = "${ VSPHERE_RESOURCE_POOL }"
+	sshAuthorizedKeysVar        = "${ SSH_AUTHORIZED_KEY }"
+	vSphereDataCenterVar        = "${ VSPHERE_DATACENTER }"
+	vSphereDatastoreVar         = "${ VSPHERE_DATASTORE }"
+	vSphereNetworkVar           = "${ VSPHERE_NETWORK }"
+	vSphereServerVar            = "${ VSPHERE_SERVER }"
+	vSphereTemplateVar          = "${ VSPHERE_TEMPLATE }"
+	defaultHAProxyTemplate      = "capv-haproxy-v0.5.3-77-g224e0ef6"
```
Comment (yastij) [inline packaging/flavorgen/flavors/generators.go]: this shouldn't be hardcoded. this needs to be provided by the user. consumers can have different names for this OVA

File: pkg/manager/constants.go [hunk 7]
```diff
@@ -23,7 +23,7 @@ const (
 
 	// DefaultWebhookServiceContainerPort is the default value for the eponymous
 	// manager option.
-	DefaultWebhookServiceContainerPort = 9443
+	DefaultWebhookServiceContainerPort = 0
```
Comment (yastij) [inline pkg/manager/constants.go]: +1

File: pkg/manager/constants.go [hunk 8]
```diff
@@ -33,17 +33,11 @@ const (
...
 	DefaultMaxConcurrentReconciles = 1
 
-	// DefaultPodNamespace is the default value for the eponymous manager
-	// option.
-	DefaultPodNamespace = defaultPrefix + "system"
-
 	// DefaultPodName is the default value for the eponymous manager option.
 	DefaultPodName = defaultPrefix + "controller-manager"
 
+	DefaultPodNamespace = defaultPrefix + "system"
```
Comment (yastij) [inline pkg/manager/constants.go]: is this is still used in the code base ?

File: pkg/manager/options.go [hunk 9]
```diff
@@ -139,4 +142,13 @@ func (o *Options) defaults() {
...
 		o.Password = os.Getenv("VSPHERE_PASSWORD")
 	}
+	if ns, ok := os.LookupEnv("POD_NAMESPACE"); ok {
+		o.PodNamespace = ns
+	} else if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
+		if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
+			o.PodNamespace = ns
+		}
+	} else {
+		o.PodNamespace = DefaultPodNamespace
```
Comment (yastij) [inline pkg/manager/options.go]: same, used ?

File: pkg/manager/constants.go (hunk 8 above)
Comment (randomvariable) [reply pkg/manager/constants.go]: yes, it goes into the event recorder. i think a follow up issue is to resolve that, and programatically find the running namespace.

File: packaging/flavorgen/flavors/flavors.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+		&machineDeployment,
+	}
+}
+
+func MultiNodeTemplateWithHAProxy() []runtime.Object {
+	lb := newHAProxyLoadBalancer()
+	vsphereCluster := newVSphereCluster(&lb)
+	cluster := newCluster(vsphereCluster)
+	machineTemplate := newVSphereMachineTemplate()
+	controlPlane := newKubeadmControlplane(444, machineTemplate)
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply packaging/flavorgen/flavors/flavors.go]: The 444 is just a integer placeholder. It gets replaced by the clusterctl variable `CONTROL_PLANE_MACHINE_COUNT` after going through unstructured processing.

File: pkg/manager/options.go (hunk 9 above)
Comment (randomvariable) [reply pkg/manager/options.go]: same as above.

follow up issue to track, wdyt?

File: Makefile [hunk 11]
```diff
@@ -52,6 +57,25 @@ WEBHOOK_ROOT ?= $(MANIFEST_ROOT)/webhook
...
 
+RELEASE_DIR := out
+BUILD_DIR := .build
+OVERRIDES_DIR := $(HOME)/.cluster-api/overrides/infrastructure-vsphere/$(VERSION)
+
+# Define Docker related variables. Releases should modify and double check these vars.
+REGISTRY ?= gcr.io/$(shell gcloud config get-value project)
+PROD_REGISTRY := gcr.io/cluster-api-provider-vsphere/release
+IMAGE_NAME ?= manager
+DEV_CONTROLLER_IMG ?= $(REGISTRY)/vsphere-$(IMAGE_NAME)
```
[the remark this answers is not in the document]
Comment (randomvariable) [reply Makefile]: what do you mean?

File: examples/default/addons.yaml (hunk 1 above)
Comment (randomvariable) [reply examples/default/addons.yaml]: clusterctl does not currently template the addons, so for ease of UX, reset back to the Calico default. Doesn't stop advanced users or downstream consumers from doing what they need to do.

File: pkg/manager/options.go (hunk 9 above)
Comment (yastij) [reply pkg/manager/options.go]: Sgtm

**Comment by randomvariable** [comment]
e2e startup modified to handle new credential management

**Pull request merged**